alibi/prototypes/protoselect.py

```diff
@@ -0,0 +1,543 @@
+
+        # update metadata
+        self.meta['params'].update({
+            'kernel_distance': kernel_distance_tag,
+            'eps': eps,
+            'lbd': lbd,
+            'batch_size': batch_size,
+            'verbose': verbose
+        })
+
+    def fit(self,
+            X: np.ndarray,
+            X_labels: Optional[np.ndarray] = None,
+            Y: Optional[np.ndarray] = None) -> 'ProtoSelect':
+        """
+        Fit the explainer by setting the reference dataset.
+
+        Parameters
+        ---------
+        X
+            Reference dataset to be summarized.
+        X_labels
+            Labels of the reference dataset.
+        Y
+            Dataset to choose the prototypes from. If ``None``, the prototypes will be selected from the reference
+            dataset `X`.
+
+        Returns
+        -------
+        self
+            Reference to itself.
+        """
+        self.X = X
+        # if the `X_labels` are not provided, then consider that all elements belong to the same class. This means
+        # that loss term which tries to avoid including in an epsilon ball elements belonging to other classes
+        # will always be 0. Still the first term of the loss tries to cover as many examples as possible with
+        # minimal overlap between the epsilon balls corresponding to the other prototypes.
+        self.X_labels = X_labels.astype(np.int32) if (X_labels is not None) else np.zeros((len(X), ), dtype=np.int32)
+        # if the set of prototypes is not provided, then find the prototypes belonging to the reference dataset.
+        self.Y = Y if (Y is not None) else self.X
+        # initialize penalty for adding a prototype
+        if self.lbd is None:
+            self.lbd = 1 / len(self.X)
+            self.meta['params'].update({'lbd': self.lbd})
+
+        self.max_label = np.max(self.X_labels)
+        self.kmatrix_yx = batch_compute_kernel_matrix(x=self.Y,
+                                                      y=self.X,
+                                                      kernel=self.kernel_distance,
+                                                      batch_size=self.batch_size,
+                                                      preprocess_fn=self.preprocess_fn)
+
+        return self
+
+    def explain(self, num_prototypes: int = 1) -> Explanation:
+        """
+        Searches for the requested number of prototypes. Note that the algorithm can return a lower number of
+        prototypes than the requested one. To increase the number of prototypes, reduce the epsilon-ball radius
+        `eps` and the penalty `lbd` for adding a prototype.
+
+        Parameters
+        ----------
+        num_prototypes
+            Number of maximum prototypes to be selected.
+
+        Returns
+        -------
+        An `Explanation` object containing the prototypes, prototypes indices and protoypes labels with additional \
+        metadata as attributes
```

Comment:
Minor grammar improvements - > "containing the prototypes, prototype indices and prototype labels..."